### DISCUSSION TITLE
Problem with inheritance and templates

### DISCUSSION THREAD

**Discussion by hoppers99** [TriliumNext/Trilium#4554] [Q&A]
I can't figure out what I'm doing wrong and I'm hoping someone can help!

I have a template called `Character Type`, which includes an attribute defined as `#label:characterType(inheritable)=promoted,single,text`.

I then use that template to create a page (called `Merrow Extortionist`) and populate the `characterType` field with a value of `Monster`. So far, we're good.

I then created a child page of `Merrow Extortionist` (called `Merrow`). What I expect to see is the attribute `characterType` shown in my promoted attributes with a value of `Monster`. Instead, I see the attribute in my promoted attributes fine, but with no value.

This is an example but I have several attributes which I want to be inherited by the child pages, all of which come through as empty labels.

My understanding from the documentation is that using inheritable in the template should cause that setting to apply to the page made from the template (rather than it being about the child of the template) and then that should apply to children made to the page the template applies. I also understand that inheritance applies to both key and value. I may be misunderstanding, however, I haven't managed to find any conversations that clarify it or discuss an issue like this.

A few screenshots to help give some more context at least:

Version info:
![image](https://github.com/zadam/trilium/assets/5240510/5ac40f42-b07a-408a-ad3d-285d2ba8c627)

Template Owned Attributes:
![image](https://github.com/zadam/trilium/assets/5240510/aa98e8bc-2481-4d9b-ae06-f92b7c71daf2)

Page made from Template (`Merrow Extortionist`) Inherited and Owned Attributes plus Promoted view:
![image](https://github.com/zadam/trilium/assets/5240510/27575127-87c1-4653-8ba2-dccf060277d2)
![image](https://github.com/zadam/trilium/assets/5240510/e660a051-b4eb-43e1-a6f0-57c35dab4810)
![image](https://github.com/zadam/trilium/assets/5240510/3c8bf6fd-26ae-4f75-bfde-c3b4f2c12c85)

Child page of above (`Merrow`) Inherited and Owned Attributes plus Promoted view:
![image](https://github.com/zadam/trilium/assets/5240510/48e7cedd-dea6-4e04-a3cc-5480d2c738d1)
![image](https://github.com/zadam/trilium/assets/5240510/ed70fe7b-fc2d-49aa-82e8-31762d7a0584)
![image](https://github.com/zadam/trilium/assets/5240510/54c1fa26-3d8b-4f4b-93c5-b89ccd3f449f)

I hope that makes sense. Happy to provide other details if there are any useful but I'm not sure enough of what's going on to know what to add. Any help/suggestions/pointers appreciated.

**Comment by rom1dep** [comment]
As I understand it (but, heh, I'm new here, so pinch of salt and everything), `characterType` is defined as a label, so marking it inheritable ensures that all children will have a label `characterType` to hold individual values, which may be different for different children (and stored for every child as the owned attribute `#characterType=theValueSetForThatChild`), but the values themselves aren't inherited (by default).

That said, apparently nothing prevents you from adding the inheritable flag manually to the value of the child, so if you set, for Merrow Extortionist, as owned attribute `#characterType(inheritable)=Monster`, Merrow should show this value in its inherited attributes. But it will still show an empty label, and can still hold/override its own value for `characterType` in such a way that you might have for Merrow `#characterType=MerrowMonster` as owned attribute and `#characterType=Monster` as inherited attribute (from Merrow Extortionist).

**Reply by rom1dep** [reply]
@zadam what could be neat in this case would be to not have an empty label as "promoted attribute", but one holding the value inherited from the parent, ideally in a lighter color (to denote it being inherited and not owned), while still letting the user override the value (i.e. making it owned). WDYT?

**Reply by zadam** [reply]
This is correct. What is being inherited is the "form input", not the value. But you can make the value inheritable as well.

**Reply by hoppers99** [reply]
Thanks @rom1dep and @zadam. What I'm trying to achieve from the template is to have it so that when I populate the value on a page that uses that template, the children of the new page will get that value inherited (but overridable of course). Is this something that can be achieved from the template at all?

In the template, I can mark the attribute itself as inheritable but that appears only to get inherited by a child of the template page, not a page using the template (hopefully that makes sense...).

The closest I've managed is to define a child: attribute in the template which is then picked up by children of the page that is created with the template, but these values are not based on the parent page values.

So do I understand correctly that I just can't template inheritance currently? That's fine if it's the case, it just means I give up on my futile mission and I'm stuck doing raw data entry for every page (or more accurately setting the top-level inheritance for every top-level page independent of the template).

**Reply by rom1dep** [reply]
> What I'm trying to achieve from the template is to have it so that when I populate the value on a page that uses that template, the children of the new page will get that value inherited (but overridable of course). Is this something that can be achieved from the template at all?

@hoppers99 : you would get this behaviour by using regular attributes (instead of labels), so the real question is, "why not doing just that"? Are labels (i.e. the "forms inputs") really that big a deal?

**Reply by hoppers99** [reply]
@rom1dep As best I can tell, the inheritance of values is only working if I mark the values as inheritable in the created page - I was trying to set in the template that these would be inheritable based on the fact the template was used. This, however, does not work even giving them seed values in the template; when the page is created with the template, the inheritable value does not carry over (the value does but not the inheritable setting) and it's also not showing that seed value in the promoted label (I haven't thought too much about that to think if that makes sense to me or not).

So every time I make a page from this template, I still need to add all the values and independently mark them as inheritable (side note: it'd be great if a promoted label could have a checkbox or something next to it to control inheritability of the value it controls).

The promoted labels were an easy way of displaying key data at a glance (perhaps I could achieve this through a rendered page instead) to go with the greater content. I think it seems I'm just trying to mix template and inheritance in a way that doesn't play nicely - I had an idea in my head about how these two could make my task a lot easier but I think I'm just going to have to rely on a more manual approach. And/or I flesh out my thinking better so I can articulate a feature request (or a PR if I find the time to understand the codebase).
